packages/models-library/src/models_library/sharing_networks.py

```diff
@@ -0,0 +1,107 @@
+from collections import deque
+from typing import Any, Deque, Dict
+from uuid import UUID
+
+from pydantic import BaseModel, constr
+
+from .projects_nodes_io import NodeID
+
+SERVICE_NETWORK_RE = r"^[a-zA-Z]([a-zA-Z0-9_-]{0,63})$"
+
+
+DockerNetworkName = constr(regex=SERVICE_NETWORK_RE)
+DockerNetworkAlias = constr(regex=SERVICE_NETWORK_RE)
+
+
+def validate_network_name(value: str) -> DockerNetworkName:
+    class ValidationModel(BaseModel):
+        item: DockerNetworkName
+
+    return ValidationModel(item=value).item
+
+
+def validate_network_alias(value: str) -> DockerNetworkAlias:
+    class ValidationModel(BaseModel):
+        item: DockerNetworkAlias
+
+    return ValidationModel(item=value).item
+
+
+class BaseModelDict(BaseModel):
+    __root__: Dict[Any, Any]
+
+    def __iter__(self):
+        return iter(self.__root__)
+
+    def __getitem__(self, item):
+        return self.__root__[item]
+
+    def __setitem__(self, item, value):
+        self.__root__[item] = value
+
+    def items(self):
+        return self.__root__.items()
+
+    def get(self, item, default=None):
+        return self.__root__.get(item, default)
+
+    @staticmethod
+    def _convert_dict_uuid_keys(dict_data: Dict[Any, Any]) -> Dict[Any, Any]:
+        to_change: Deque[UUID] = deque()
+        for key in dict_data.keys():
+            if isinstance(key, UUID):
+                to_change.append(key)
+
+        for key in to_change:
+            dict_data[f"{key}"] = dict_data.pop(key)
+
+        return dict_data
+
+    def _iter(self, *args, **kwargs) -> "TupleGenerator":
+        # ensure dict key conversion works with either
+        # .json() and .dict() methods
+        for tuple_generator in super()._iter(*args, **kwargs):
+            dict_key, value = tuple_generator
+            if isinstance(value, dict):
+                value = self._convert_dict_uuid_keys(value)
+            yield dict_key, value
+
+    def dict(self, *args, **kwargs) -> Dict[str, Any]:
+        return super().dict(*args, **kwargs)["__root__"]
+
+
+class ContainerAliases(BaseModelDict):
+    __root__: Dict[NodeID, DockerNetworkAlias]
+
+
+class SharingNetworks(BaseModelDict):
+    __root__: Dict[DockerNetworkName, ContainerAliases]
+
+    class Config:
+        schema_extra = {
+            "examples": [
+                {"nSetwork_name12-s": {"5057e2c1-d392-4d31-b5c8-19f3db780390": "ok"}},
+                {"C": {"5057e2c1-d392-4d31-b5c8-19f3db780390": "ok"}},
+                {"shr-ntwrk_5c743ad2-8fdb-11ec-bb3a-02420a000008_default": {}},
```

Comment:
not sure I get that. how do you connect with the node uuid?

Reply:
A project will typically have more than one network, looking something like the following:

```json
{
    "shr-ntwrk_5c743ad2-8fdb-11ec-bb3a-02420a000008_jupyters": {
        "00000000-0000-0000-0000-000000000000": "network_alias_jupyter_1",
        "00000000-0000-0000-0000-000000000001": "network_alias_jupyter_2",
    },
    "shr-ntwrk_5c743ad2-8fdb-11ec-bb3a-02420a000008_smashes": {
        "00000000-0000-0000-0000-000000000002": "network_alias_smashe_1",
        "00000000-0000-0000-0000-000000000003": "network_alias_smashe_1",
    },
}
```

The keywords `jupyters` and `smashes` from the network name are meant to be defined by the user. In this iteration we use only 1 network with the keyword `default`.
The network aliases `network_alias_jupyter_1`, `network_alias_jupyter_2`, `network_alias_smashe_1` and `network_alias_smashe_2` are also meant  to be defined by the user. In this iteration they are procured from the label assigned in the UI. If this label is not transformable in a valid network alias, it will not be added.

Reply:
I do not understand your point. In the PR description above you define 1 network per study, which is what we agreed. Now you create several ones. Why? What does this bring? 
And for the MVP I am not sure we need to modify the project data structure, or do we?

Reply:
What I illustrate above will be required in the future. When we have a UI which will allow the user to specify groups of nodes which are connected together. What this bring is a common implementation that can easily be extended in the future, but just adding a few API calls for the frontend.

In this current implementation the sharing_networks object will look like below

```json
{
    "shr-ntwrk_00000000-0000-0000-0000-000000000000_default": {
        "00000000-0000-0000-0000-000000000000": "network_alias_jupyter_1",
        "00000000-0000-0000-0000-000000000001": "network_alias_jupyter_2",
        "00000000-0000-0000-0000-000000000002": "network_alias_smashe_1",
        "00000000-0000-0000-0000-000000000003": "network_alias_smashe_1"
    }
}
```

Meaning all the dynamic-sidecar nodes are placed on the same "default" network.

Regarding the storing of the data. This decision was taken since the idea behind this PR was to enable in the future an easy way to attach the a GUI. Since that required multiple networks with user defined groups of nodes, it is required to store it somewhere. The project looked like the most promising place to keep this data.

Reply:
This was already moved outside the project. Sorry the comment got a bit old.